src/_image_wrapper.cpp

```diff
@@ -49,249 +50,178 @@ const char* image_resample__doc__ =
-    PyArrayObject *output = NULL;
-    PyArrayObject *transform_mesh = NULL;
-    int ndim;
-    int type;
-
-    params.interpolation = NEAREST;
-    params.transform_mesh = NULL;
-    params.resample = false;
-    params.norm = false;
-    params.radius = 1.0;
-    params.alpha = 1.0;
-
-    const char *kwlist[] = {
-        "input_array", "output_array", "transform", "interpolation",
-        "resample", "alpha", "norm", "radius", NULL };
-
-    if (!PyArg_ParseTupleAndKeywords(
-            args, kwargs, "OOO|iO&dO&d:resample", (char **)kwlist,
-            &py_input, &py_output, &py_transform,
-            &params.interpolation, &convert_bool, &params.resample,
-            &params.alpha, &convert_bool, &params.norm, &params.radius)) {
-        return NULL;
-    }
+    if (ndim < 2 || ndim > 3)
+        throw std::invalid_argument("Input array must be a 2D or 3D array");
 
-    if (params.interpolation < 0 || params.interpolation >= _n_interpolation) {
-        PyErr_Format(PyExc_ValueError, "Invalid interpolation value %d",
-                     params.interpolation);
-        goto error;
-    }
+    if (ndim == 3 && input_array.shape(2) != 4)
+        throw std::invalid_argument(
+            "3D input array must be RGBA with shape (M, N, 4), has trailing dimension of " +
+            std::to_string(ndim));
 
-    input = (PyArrayObject *)PyArray_FromAny(
-        py_input, NULL, 2, 3, NPY_ARRAY_C_CONTIGUOUS, NULL);
-    if (!input) {
-        goto error;
-    }
-    ndim = PyArray_NDIM(input);
-    type = PyArray_TYPE(input);
+    // Ensure input array is contiguous, regardless of dtype
+    input_array = pybind11::array::ensure(input_array, pybind11::array::c_style);
 
-    if (!PyArray_Check(py_output)) {
-        PyErr_SetString(PyExc_ValueError, "Output array must be a NumPy array");
-        goto error;
-    }
-    output = (PyArrayObject *)py_output;
-    if (PyArray_NDIM(output) != ndim) {
-        PyErr_Format(
-            PyExc_ValueError,
-            "Input (%dD) and output (%dD) have different dimensionalities.",
-            ndim, PyArray_NDIM(output));
-        goto error;
-    }
-    // PyArray_FromAny above checks that input is 2D or 3D.
-    if (ndim == 3 && (PyArray_DIM(input, 2) != 4 || PyArray_DIM(output, 2) != 4)) {
-        PyErr_Format(
-            PyExc_ValueError,
-            "If 3D, input and output arrays must be RGBA with shape (M, N, 4); "
-            "got trailing dimensions of %" NPY_INTP_FMT " and %" NPY_INTP_FMT
-            " respectively", PyArray_DIM(input, 2), PyArray_DIM(output, 2));
-        goto error;
-    }
-    if (PyArray_TYPE(output) != type) {
-        PyErr_SetString(PyExc_ValueError, "Mismatched types");
-        goto error;
-    }
-    if (!PyArray_IS_C_CONTIGUOUS(output)) {
-        PyErr_SetString(PyExc_ValueError, "Output array must be C-contiguous");
-        goto error;
-    }
+    // Validate output array
+    auto out_ndim = output_array.ndim();
+
+    if (out_ndim != ndim)
```

Comment:
Although this works, I prefer having curly braces, even when there is a single line.

(If this is consistent with matplotlib's style, you are welcome to ignore this commit).

Reply:
Ignoring :smile:

Reply:
Curly braces added around all one-liners.